### PR TITLE
perf(ops): trim async op cruft

### PR DESCRIPTION
**Pull request by mmastrac**
Removes a lot of the old async machinery that is no longer used, and provides a path forward for op tracking, used in Deno test. The metrics object has been trimmed down significantly.

Before:
```
test bench_op_async_void                  ... bench:     104,595 ns/iter (+/- 4,252)
```

After:
```
test bench_op_async_void                  ... bench:      75,782 ns/iter (+/- 3,128)
```

Changes:

 - Don't create a scope for async ops unless or until we really need one
 - Start using the new metrics functions
 - Replace OpTracker with a new metrics tracker that is optional and zero-cost when unused

### CODE REVIEW

File: core/ops.rs [hunk 1]
```diff
@@ -177,9 +158,9 @@ impl OpCtx {
     // version and require a slightly different set of arguments (for example, it may need the fastcall
     // callback information to get the `OpCtx`).
     let fast_fn = if metrics_fn.is_some() {
-      &decl.fast_fn
-    } else {
       &decl.fast_fn_with_metrics
+    } else {
+      &decl.fast_fn
```
Comment (bartlomieju) [inline core/ops.rs:163]: Ooops, a previous mistake?

File: core/ops_metrics.rs [hunk 2]
```diff
@@ -1,92 +1,136 @@
...
-pub struct OpMetrics {
-  pub ops_dispatched: u64,
+pub struct OpMetricsSummary {
   pub ops_dispatched_sync: u64,
   pub ops_dispatched_async: u64,
-  // TODO(bartlomieju): this field is never updated
-  pub ops_dispatched_async_unref: u64,
-  pub ops_completed: u64,
-  pub ops_completed_sync: u64,
   pub ops_completed_async: u64,
```
Comment (bartlomieju) [inline core/ops_metrics.rs:60]: We should make sure the shape of the object returned from `Deno.metrics()` API stays the same - the fields can be zeroed out, but the shape should still be correct until the API is present on the `Deno` namespace.

File: core/ops_metrics.rs [hunk 3]
```diff
@@ -1,92 +1,136 @@
...
 
-// TODO(@AaronO): track errors
-#[derive(Default, Debug)]
-pub struct OpsTracker {
-  ops: RefCell<Vec<OpMetrics>>,
+impl OpMetricsSummary {
+  /// Does this op have outstanding async op dispatches?
+  pub fn has_outstanding_ops(&self) -> bool {
+    self.ops_dispatched_async > self.ops_completed_async
+  }
```
Comment (bartlomieju) [inline core/ops_metrics.rs:67]: Doesn't it need to take ref/unref ops into account?

File: core/runtime/ops.rs [hunk 4]
```diff
@@ -26,117 +23,6 @@ use v8::WriteOptions;
...
-  match op {
-    Err(err) => Either::Left(ready(OpResult::Err(OpError::new(
-      ctx.get_error_class_fn,
-      err,
-    )))),
-    Ok(fut) => Either::Right(fut.map(|r| OpResult::Ok(r.into()))),
-  }
-}
-
-pub fn queue_async_op<'s>(
```
Comment (bartlomieju) [inline core/runtime/ops.rs:97]: Lovely to see all of these go!

**Review by bartlomieju — APPROVED** [review]

File: core/ops.rs (hunk 1 above)
Comment (mmastrac) [reply core/ops.rs:163]: Yeah 😓  We never actually enabled metrics in fast ops. I added the fast ops to the benchmark so it'll have some coverage.

File: core/ops_metrics.rs (hunk 3 above)
Comment (mmastrac) [reply core/ops_metrics.rs:67]: It never did in the past.. those were always zero 🤷

**Pull request merged**